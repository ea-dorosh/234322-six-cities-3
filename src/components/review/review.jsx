import React, {PureComponent} from "react";
import PropTypes from "prop-types";
import {ratingToStar} from "../../utils.js";


class Review extends PureComponent {
  constructor(props) {
    super(props);
    this.dateStamp = this.dateStamp.bind(this);
  }

  dateStamp(date) {
    const dateObj = new Date(date);

    const MONTHS = [`January`, `February`, `March`, `April`, `May`, `June`, `July`, `August`, `September`, `October`, `November`, `December`];

    return `${MONTHS[dateObj.getMonth()]} ${dateObj.getFullYear()}`;
  }

  render() {
    const {rating, name, text, date} = this.props.review;

    // eslint-disable-next-line react/prop-types
    const ratingReview = ratingToStar(rating);

    return (
      <li className="reviews__item">
        <div className="reviews__user user">
          <div className="reviews__avatar-wrapper user__avatar-wrapper">
            <img className="reviews__avatar user__avatar" src="/img/avatar-max.jpg" width="54" height="54"
              alt="Reviews avatar"/>
          </div>
          <span className="reviews__user-name">
            {name}
          </span>
        </div>
        <div className="reviews__info">
          <div className="reviews__rating rating">
            <div className="reviews__stars rating__stars">
              <span style={{width: ratingReview + `%`}}/>
              <span className="visually-hidden">Rating</span>
            </div>
          </div>
          <p className="reviews__text">
            {text}
          </p>
          <time className="reviews__time" dateTime={date}>{this.dateStamp(date)}</time>
        </div>
      </li>
    );
  }
}

Review.propTypes = {
  review: PropTypes.shape(
      {
        rating: PropTypes.number.isRequired,
        name: PropTypes.string.isRequired,
        text: PropTypes.string.isRequired,
        date: PropTypes.string.isRequired,
      }
  )
};


export default Review;

