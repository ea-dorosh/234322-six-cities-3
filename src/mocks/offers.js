import {getRandomNumber, getRandomElement} from "../utils";

const PRICE = {
  MIN: 50,
  MAX: 500
};

const offersNames = [
  `Beautiful & luxurious apartment at great location`,
  `Nice, cozy, wood and stone place`,
  `Wood and stone place`,
  `Nice, cozy, warm big bed apartment`
];

const holderNames = [
  `Angelina`,
  `Max`,
];

const offersTypes = [
  `apartment`,
  `room`,
  `house`,
  `hotel`
];

const offersCords = [
  [52.3909553943508, 4.85309666406198],
  [52.369553943508, 4.85309666406198],
  [52.3909553943508, 4.929309666406198],
  [52.3809553943508, 4.939309666406198]
];

const createOffer = (index) => {
  const bedRoomQuantity = getRandomNumber(1, 6);
  const maxGuestQuantity = bedRoomQuantity * 2;
  const holderName = getRandomElement(holderNames);

  return {
    img: `img/apartment-0${getRandomNumber(1, 3)}.jpg`,
    photos: [
      `img/apartment-0${getRandomNumber(1, 3)}.jpg`,
      `img/apartment-0${getRandomNumber(1, 3)}.jpg`,
      `img/apartment-0${getRandomNumber(1, 3)}.jpg`,
      `img/apartment-0${getRandomNumber(1, 3)}.jpg`,
      `img/apartment-0${getRandomNumber(1, 3)}.jpg`,
      `img/apartment-0${getRandomNumber(1, 3)}.jpg`,
    ],
    isPremium: !!(index % 2),
    price: getRandomNumber(PRICE.MIN, PRICE.MAX),
    name: offersNames[index],
    type: getRandomElement(offersTypes),
    rating: parseFloat(`${getRandomNumber(2, 4)}.${getRandomNumber(0, 9)}`),
    description: `An independent House, strategically located between Rembrand Square and National Opera, but where the bustle of the city comes to rest in this alley flowery and colorful.`,
    bedRoomQuantity,
    maxGuestQuantity,
    details: [
      `Wi-Fi`,
      `Washing machine`,
      `Towels`,
      `Heating`,
      `Coffee machine`,
      `Baby seat`,
      `Kitchen`,
      `Dishwasher`,
      `Cabel TV`,
      `Fridge`,
    ],
    holder: {
      holderName,
      img: `img/avatar-${holderName.toLowerCase()}.jpg`,
      isSuper: !(index % 2),
    },
    coords: {
      x: offersCords[index][0],
      y: offersCords[index][1],
    }
  };
};

const createOffers = (amount) => {
  const offers = [];

  for (let i = 0; i < amount; i++) {
    const offer = createOffer(i);

    offers.push(offer);
  }
  return offers;

};


export {createOffers};
