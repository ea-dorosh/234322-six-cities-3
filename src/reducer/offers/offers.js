import {extend, getCities} from "../../utils.js";


const initialState = {
  load: null,
  error: null,
  cities: [],
  activeCity: null,
  offers: [],
};

const ActionType = {
  LOAD_OFFERS: `LOAD_OFFERS`,
  CHANGE_CITY: `CHANGE_CITY`,
  GET_ERROR: `GET_ERROR`
};

const ActionCreator = {
  loadOffers: (offers) => ({
    type: ActionType.LOAD_OFFERS,
    payload: offers
  }),
  changeCity: (city) => ({
    type: ActionType.CHANGE_CITY,
    payload: city,
  }),
  getError: (err) => ({
    type: ActionType.GET_ERROR,
    payload: err,
  })
};

const Operation = {
  download: () => (dispatch, getState, api) => {
    return api.get(`/hotels`)
      .then((response) => {
        dispatch(ActionCreator.loadOffers(response.data));
      })
      .catch((err) => {
        throw err;
      });
  }
};

const reducer = (state = initialState, action) => {
  switch (action.type) {
    case ActionType.LOAD_OFFERS:
      const cities = getCities(action.payload);

      return extend(state, {
        load: true,
        cities,
        offers: action.payload,
        activeCity: cities[0]
      });

    case ActionType.CHANGE_CITY:
      return extend(state, {activeCity: action.payload});

    case ActionType.GET_ERROR:
      return extend(state, {error: action.payload, load: `fail`});
  }

  return state;
};

export {reducer, Operation, ActionType, ActionCreator};
