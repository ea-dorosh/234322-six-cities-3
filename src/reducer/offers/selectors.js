import NameSpace from "../name-space.js";
import {SortType} from "../../utils";
import {createSelector} from "reselect";
import {getSortType} from "../main/selectors";
import {getOffersByCity} from "../../utils";


const NAME_SPACE = NameSpace.OFFERS;

export const getLoadStatus = (state) => {
  return state[NAME_SPACE].loadingOfferStatus;
};

export const getNearOffers = (state) => {
  return state[NAME_SPACE].nearOffers;
};

export const getError = (state) => {
  return state[NAME_SPACE].error;
};

export const getActiveCity = (state) => {
  return state[NAME_SPACE].activeCity;
};

export const getCitiesList = (state) => {
  return state[NAME_SPACE].cities;
};

function getOffers(state) {
  return state[NAME_SPACE].offers;
}

const getOffersByActiveCity = createSelector([
  getOffers,
  getActiveCity,
], (offers, activeCity) => getOffersByCity(offers, activeCity.name)
);

function getSortedOffers(offers, sortType) {
  switch (sortType) {
    case SortType.PRICE_TO_LOW:
      return offers.slice().sort((a, b) => b.price - a.price);
    case SortType.PRICE_TO_HIGH:
      return offers.slice().sort((a, b) => a.price - b.price);
    case SortType.TOP_RATED:
      return offers.slice().sort((a, b) => b.rating - a.rating);
  }
  return offers;
}


export const sortOffersBySortType = createSelector([
  getOffersByActiveCity,
  getSortType,
], (offers, sortType) => getSortedOffers(offers, sortType)
);

export const selectOffers = (state) => {
  return sortOffersBySortType(state);
};
