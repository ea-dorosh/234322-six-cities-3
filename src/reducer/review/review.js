import {extend, prepareReview} from '../../utils.js';

const LoadingStatus = {
  PROCESSING: `PROCESSING`,
  SUCCESS: `SUCCESS`,
  FAILED: `FAILED`,
};

const initialState = {
  loadingStatus: LoadingStatus.SUCCESS,
  sendStatus: ``,
  offerReviews: [],
};

const ActionType = {
  CHANGE_LOADING_STATUS: `CHANGE_LOADING_STATUS`,
  CHANGE_SENDING_STATUS: `CHANGE_SENDING_STATUS`,
  GET_OFFER_REVIEWS: `GET_OFFER_REVIEWS`,
};

const ActionCreator = {
  changeLoadingStatus: (status) => ({
    type: ActionType.CHANGE_LOADING_STATUS,
    payload: status
  }),
  changeSendingStatus: (status) => ({
    type: ActionType.CHANGE_SENDING_STATUS,
    payload: status
  }),
  getOfferReviews: (data) => ({
    type: ActionType.GET_OFFER_REVIEWS,
    payload: (data)
  })
};

export const Operation = {
  postReview: (reviewData, id) => (dispatch, getState, api) => {
    dispatch(ActionCreator.changeSendingStatus(LoadingStatus.PROCESSING));
    return api.post(`/comments/${id}`, reviewData)
      .then(() => {
        dispatch(ActionCreator.changeSendingStatus(LoadingStatus.SUCCESS));
        dispatch(Operation.getReviews(id));
      })
      .catch(() => {
        dispatch(ActionCreator.changeSendingStatus(LoadingStatus.FAILED));
      });
  },
  getReviews: (id) => (dispatch, getState, api) => {
    dispatch(ActionCreator.changeLoadingStatus(LoadingStatus.PROCESSING));
    return api.get(`/comments/${id}`)
      .then((response) => {
        const reviewsData = prepareReview(response.data);
        dispatch(ActionCreator.getOfferReviews(reviewsData));
      })
      .catch(() => {
        dispatch(ActionCreator.changeLoadingStatus(LoadingStatus.FAILED));
      });
  }
};

const reducer = (state = initialState, action) => {
  switch (action.type) {
    case ActionType.CHANGE_LOADING_STATUS:
      return extend(state, {
        loadingStatus: action.payload
      });

    case ActionType.GET_OFFER_REVIEWS:
      return extend(state, {
        offerReviews: action.payload,
        loadingStatus: LoadingStatus.SUCCESS,
      });

    case ActionType.CHANGE_SENDING_STATUS:
      return extend(state, {
        sendStatus: action.payload
      });
  }

  return state;
};

export {reducer, ActionType, ActionCreator, LoadingStatus};
