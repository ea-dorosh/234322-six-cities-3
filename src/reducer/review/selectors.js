import NameSpace from '../name-space.js';

export const getLoadingStatus = (state) => state[NameSpace.REVIEW].loadingStatus;
