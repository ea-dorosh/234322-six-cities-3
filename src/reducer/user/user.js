import {extend} from "../../utils.js";

const AuthorizationStatus = {
  AUTH: `AUTH`,
  NO_AUTH: `NO_AUTH`,
};

const initialState = {
  authorizationStatus: AuthorizationStatus.NO_AUTH,
  userProperties: null,
};

const ActionType = {
  REQUIRED_AUTHORIZATION: `REQUIRED_AUTHORIZATION`,
  GET_USER_PROPERTIES: `GET_USER_PROPERTIES`,
};

const ActionCreator = {
  requireAuthorization: (status) => {
    return {
      type: ActionType.REQUIRED_AUTHORIZATION,
      payload: status,
    };
  },
  getUserProperties: (data) => {
    return {
      type: ActionType.GET_USER_PROPERTIES,
      payload: data,
    };
  },
};

const Operation = {
  checkAuth: () => (dispatch, getState, api) => {
    return api.get(`/login`)
      .then((response) => {
        dispatch(ActionCreator.requireAuthorization(AuthorizationStatus.AUTH));
        dispatch(ActionCreator.getUserProperties(response.data));
      })
      .catch(() => {
        // eslint-disable-next-line no-console
        console.log(`вы не авторизованы`);
      });
  },

  login: (authData) => (dispatch, getState, api) => {
    return api.post(`/login`, {
      email: authData.login,
      password: authData.password,
    })
      .then((response) => {
        dispatch(ActionCreator.requireAuthorization(AuthorizationStatus.AUTH));
        dispatch(ActionCreator.getUserProperties(response.data));
      });
  },
};

const reducer = (state = initialState, action) => {
  switch (action.type) {
    case ActionType.REQUIRED_AUTHORIZATION:
      return extend(state, {authorizationStatus: action.payload});

    case ActionType.GET_USER_PROPERTIES:
      return extend(state, {userProperties: action.payload});
  }

  return state;
};


export {
  ActionCreator,
  ActionType,
  AuthorizationStatus,
  Operation,
  reducer,
};
