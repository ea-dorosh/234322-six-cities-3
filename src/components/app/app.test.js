import React from "react";
import renderer from "react-test-renderer";
import {Provider} from "react-redux";
import configureStore from "redux-mock-store";
import NameSpace from "../../reducer/name-space.js";
import {AuthorizationStatus} from "../../reducer/user/user.js";
import {LoadingOfferStatus} from "../../reducer/offers/offers";
import {LoadingStatus} from "../../reducer/review/review";
import App from "./app";

const mockStore = configureStore([]);

it(`Should FavoritesCityItem render correctly`, () => {
  const store = mockStore({
    [NameSpace.OFFERS]: {
      loadingOfferStatus: LoadingOfferStatus.PROCESSING,
      error: null,
      cities: [],
      activeCity: null,
      offers: [],
      nearOffers: [],
    },
    [NameSpace.USER]: {
      authorizationStatus: AuthorizationStatus.NO_AUTH,
      userProperties: null,
    },
    [NameSpace.MAIN]: {
      sortType: `Popular`,
      marker: null,
    },
    [NameSpace.FAVORITE]: {
      favoriteStatus: ``,
    },
    [NameSpace.REVIEW]: {
      loadingStatus: LoadingStatus.SUCCESS,
      sendStatus: ``,
      offerReviews: [],
    },
  });
  const tree = renderer
    .create(
        <Provider store={store}>
          <App/>
        </Provider>
    )
    .toJSON();

  expect(tree).toMatchSnapshot();
});
