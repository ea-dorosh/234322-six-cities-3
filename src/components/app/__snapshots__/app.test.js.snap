// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should FavoritesCityItem render correctly 1`] = `
<div
  style={
    Object {
      "fontSize": "60px",
    }
  }
>
  <p>
    Connection...
  </p>
</div>
`;
